### PR TITLE
Fork off: child support

### PR DESCRIPTION
**Pull request by krzysztofziobro**
# Description

Adding support for child tries to fork-off. This should make some basic contracts usable on the new chain (anything that depends on block numbers is still broken).

## Type of change

- Bug fix (?)
- New feature (?)

# Checklist:
- I have made corresponding changes to the existing documentation

### CODE REVIEW

**Comment by deuszx** [comment]
I have to admit, without knowledge about internal design of the storage it's difficult to review for the correctness/completeness. Can we add e2e tests for this new feature?

**Comment by krzysztofziobro** [comment]
> I have to admit, without knowledge about internal design of the storage it's difficult to review for the correctness/completeness. Can we add e2e tests for this new feature?

Right, I may add some tests for contract interaction (or something else that uses child tries, but I am not aware off any other useful pallets that use them) on the "forked-off" chain, but there are no e2e tests for fork-off as for now, so the setup will probably take a while.

**Review by DamianStraszak — CHANGES_REQUESTED** [review]

File: fork-off/src/jsonrpc_client.rs [hunk 1]
```diff
@@ -28,6 +28,22 @@ pub trait Rpc {
         start_key: Option<StorageKey>,
         at: Option<BlockHash>,
     ) -> Result<Vec<StorageKey>, Error>;
+
+    #[rpc(name = "childstate_getKeys")]
```
[the remark this answers is not in the document]
Comment (krzysztofziobro) [reply fork-off/src/jsonrpc_client.rs]: changed

File: fork-off/src/main.rs [hunk 2]
```diff
@@ -56,9 +56,12 @@ async fn main() -> anyhow::Result<()> {
...
 
-    let initial_state: Storage =
-        serde_json::from_value(initial_spec["genesis"]["raw"]["top"].take())
-            .expect("Deserialization of state from given chainspec file failed");
+    // Initialize with state from chainspec + empty child storage
+    let initial_state = Storage {
+        top: serde_json::from_value(initial_spec["genesis"]["raw"]["top"].take())
+            .expect("Deserialization of state from given chainspec file failed"),
+        child_storage: ChildStorage::new(),
+    };
```
Comment (deuszx) [inline fork-off/src/main.rs:64]: We can create a constructor for `Storage` that accepts the `initial_spec` only:
```rust
impl Storage {
  pub fn new(initial_spec) -> Result<Self, anyhow::Error> {
    ...
  }
}
```

Comment (krzysztofziobro) [reply fork-off/src/main.rs]: done

File: fork-off/src/fetching.rs [hunk 3]
```diff
@@ -37,10 +37,22 @@ impl StateFetcher {
...
             let mut output_guard = output.lock();
-            output_guard.insert(key, value);
-            if output_guard.len() % LOG_PROGRESS_FREQUENCY == 0 {
-                info!("Fetched {} values", output_guard.len());
+            output_guard.top.insert(key.clone(), value);
+            if let Ok(child_storage_map) = child_storage_map_res {
+                info!("Fetched child trie with {} keys", child_storage_map.len());
+                output_guard
+                    .child_storage
+                    .insert(key.without_child_storage_prefix(), child_storage_map);
```
Comment (deuszx) [inline fork-off/src/fetching.rs:51]: We can omit these if `child_storage_map` is empty.

File: fork-off/src/fetching.rs [hunk 4]
```diff
@@ -37,10 +37,22 @@ impl StateFetcher {
                 .await
                 .unwrap();
 
+            let child_storage_map_res = self
+                .client
+                .get_child_storage_for_key(key.clone(), &block)
+                .await;
```
Comment (deuszx) [inline fork-off/src/fetching.rs:43]: Above if `self.client.get_storage` returns error (`RpcError`) we call `unwrap()` and short-circuit the loop.

File: fork-off/src/jsonrpc_client.rs [hunk 5]
```diff
@@ -71,6 +89,46 @@ impl Client {
...
+                    empty_prefix.clone(),
+                    CHUNK_SIZE,
+                    start_key,
+                    Some(at.clone()),
+                )
+                .await?;
+
+            let values = self
+                .client
+                .get_child_storage_entries(child_key.clone(), keys.clone(), Some(at.clone()))
```
Comment (deuszx) [inline fork-off/src/jsonrpc_client.rs:116]: I'm wondering - the `get_child_storage_entries` might be called with up to `CHUNK_SIZE` keys, can it happen that the amount of data returned here would be too much and the query would fail?

File: fork-off/src/fetching.rs (hunk 3 above)
Comment (krzysztofziobro) [reply fork-off/src/fetching.rs:51]: Here success of the rpc means that this is indeed a key which is used to store a child trie, so we probably want to insert that anyway.

File: fork-off/src/fetching.rs (hunk 4 above)
Comment (krzysztofziobro) [reply fork-off/src/fetching.rs:43]: Error here means that we passed an incorrect storage key: since it is a correct key in general, we know that there is just no child trie here (now I realize that we actually do not know that - error can probably have a different cause, will check that)

File: fork-off/src/fetching.rs (hunk 3 above)
Comment (deuszx) [reply fork-off/src/fetching.rs:51]: I see, so we would insert empy storage which is an information as well.

File: fork-off/src/fetching.rs (hunk 4 above)
Comment (deuszx) [reply fork-off/src/fetching.rs:43]: If there is no child trie there we should be able to recognize that differently. Maybe change the return type to `Result<Option<_>, RpcError>` and handle it on the caller's site adequately.

File: fork-off/src/jsonrpc_client.rs (hunk 5 above)
Comment (krzysztofziobro) [reply fork-off/src/jsonrpc_client.rs:116]: Hmm... it's still hard bound by `CHUNK_SIZE` * (what can fit in a block), but yeah, I'm not sure if that's good enough, will change it just to be sure

File: fork-off/src/fetching.rs (hunk 4 above)
Comment (krzysztofziobro) [reply fork-off/src/fetching.rs:43]: Sounds right

File: fork-off/src/jsonrpc_client.rs (hunk 5 above)
Comment (deuszx) [reply fork-off/src/jsonrpc_client.rs:116]: When you say _what can fit in a block_ do you mean block's maximum size (in bytes) or its maximum gas? The reason I'm asking is that, in theory, you can have a very small transaction that will write lots of keys (child storage entries?) - just write new keys in a loop until you run out of gas.

Of course, this is not very practical example but there are other cases where a simple transaction can write a lot of keys.

On the other hand, maybe it's not worthwhile to spend too much time on this since it's a testing tool that won't affect any running system and we can fix as we go and encounter this problem.

**tl;dr;**: I think I'm fine leaving this _as-is_ and deal with it if we ever have to.

File: fork-off/src/fetching.rs (hunk 4 above)
Comment (deuszx) [reply fork-off/src/fetching.rs:43]: If you do that, we can `unwrap()` here as well (or just `.await?`) since the only errors that would be returned are RPC errors (connection etc.) since the _child storage does not exist_ (valid case) would be represented as `None` and be handled properly below.

File: fork-off/src/jsonrpc_client.rs (hunk 5 above)
Comment (krzysztofziobro) [reply fork-off/src/jsonrpc_client.rs:116]: I mean size, it's not possible to write more data under a key than what fits in a block (right?)
Not sure how a lots of keys can hurt us here - we have 1:1 for (keys in child trie) : (child trie storage entries)

Comment (krzysztofziobro) [reply fork-off/src/jsonrpc_client.rs:116]: Right, right, for some reason I didn't consider the case in which transaction is small but runtime/some contract generates large "value" to store it under some key

Comment (deuszx) [reply fork-off/src/jsonrpc_client.rs:116]: But again, I don't think we _have to_ handle that case here right away. I think it will be most instructive to simply run this on already-existing networks and see whether it works as expected and handle the problems as they go.

File: fork-off/src/jsonrpc_client.rs [hunk 6]
```diff
@@ -71,6 +89,65 @@ impl Client {
...
+            Err(RpcError::JsonRpcError(err)) => {
+                // Ugly :(
+                if err.message == "Client error: Invalid child storage key" {
+                    Ok(None)
+                } else {
+                    Err(RpcError::JsonRpcError(err))
+                }
+            }
+            _ => res.map(Some),
+        }
```
[the remark this answers is not in the document]
Comment (krzysztofziobro) [reply fork-off/src/jsonrpc_client.rs]: changed

[the remark this answers is not in the document]
Comment (deuszx) [reply fork-off/src/jsonrpc_client.rs]: Thanks

**Review by DamianStraszak — APPROVED** [review]

File: fork-off/src/fetching.rs (hunk 4 above)
Comment (krzysztofziobro) [reply fork-off/src/fetching.rs]: I think it's done

Comment (deuszx) [reply fork-off/src/fetching.rs]: Yes, it is. If @DamianStraszak approved I guess it means that works :)

**Review by deuszx — APPROVED** [review]
